src/interfaces/message-reader.ts

```diff
@@ -0,0 +1,79 @@
+/**
+ * A Class that consumes a Buffer stepwise,
+ * while constructing an Array with the extracted and decoded contents
+ */
+export class MessageReader {
+    private buffer: Buffer;
+    private result = [];
+
+    /**
+     * @param packet
+     *      The ENetPacket to extract data from
+     */
+    constructor(packet: any) {
+        this.buffer = packet.data();
+    }
+
+    /**
+     * Reads an integer from the buffer
+     */
+    public readInt() {
+        const c = this.readNext();
+
+        if (c === 0x80) {
+            this.result.push(
+                this.readNextN(2).readInt16LE(0)
+            );
+        } else if (c === 0x81) {
+            this.result.push(
+                this.readNextN(4).readInt32LE(0)
+            );
+        } else {
+            this.result.push(c);
+        }
+
+        return this;
+    }
+
+    /**
+     * Reads a string from the buffer
+     */
+    public readString() {
+        let char;
+        let chars = [];
+
+        do {
+            char = this.readNext();
+            chars.push(char);
+        } while(char !== 0);
+
+        this.result.push(
+            // drop null byte at the end and convert to string
+            Buffer.from(chars.slice(0, -1)).toString()
+        );
+
+        return this;
```

Comment:
Not 100% sure about the Builder pattern yet... if the structure of a message depends on its contents as well, this might turn out to be uglier than just returning the values one-by-one.

Reply:
Guess we'll just have to wait and see, huh?